runtime/gc_glue_java/RealtimeGCDelegate.hpp

```diff
@@ -0,0 +1,125 @@
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(REALTIMEGCDELEGATE_HPP_)
+#define REALTIMEGCDELEGATE_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+
+#if defined(J9VM_GC_REALTIME)
+
+#include "omr.h"
+#include "BaseNonVirtual.hpp"
+#include "EnvironmentRealtime.hpp"
+#include "GCExtensions.hpp"
+
+class MM_RealtimeGCDelegate : public MM_BaseNonVirtual
+{
+private:
+	OMR_VM *_vm;
+	MM_GCExtensions *_extensions;
+	MM_RealtimeGC *_realtimeGC;
+	J9JavaVM *_javaVM;
+
+public:
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+ 	bool _unmarkedImpliesClasses; /**< if true the mark bit can be used to check is class alive or not */
+#endif /* defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING) */
+	
+#if defined(J9VM_GC_FINALIZATION)	
+	bool _finalizationRequired;
+#endif /* J9VM_GC_FINALIZATION */
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+	bool _dynamicClassUnloadingEnabled;
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+
+	MM_RealtimeGCDelegate(MM_EnvironmentBase *env, MM_RealtimeGC *realtimeGC) :
+		_vm(env->getOmrVM()),
+		_extensions(MM_GCExtensions::getExtensions(env)),
+		_realtimeGC(realtimeGC),
+		_javaVM((J9JavaVM*)env->getOmrVM()->_language_vm) {}
+
+	bool initialize(MM_EnvironmentBase *env);
+	void tearDown(MM_EnvironmentBase *env);
+
+	bool allocateAndInitializeReferenceObjectLists(MM_EnvironmentBase *env);
+	bool allocateAndInitializeUnfinalizedObjectLists(MM_EnvironmentBase *env);
+	bool allocateAndInitializeOwnableSynchronizerObjectLists(MM_EnvironmentBase *env);
+
+#if defined(J9VM_GC_FINALIZATION)	
+	bool isFinalizationRequired() { return _finalizationRequired; }
+#endif /* J9VM_GC_FINALIZATION */
+
+	void masterSetupForGC(MM_EnvironmentBase *env);
+	void masterCleanupAfterGC(MM_EnvironmentBase *env);
+	void incrementalCollect(MM_EnvironmentRealtime *env);
+	void doAuxilaryGCWork(MM_EnvironmentBase *env);
+	void clearGCStats(MM_EnvironmentBase *env);
+	void reportSyncGCEnd(MM_EnvironmentBase *env);
+
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+	MMINLINE bool isDynamicClassUnloadingEnabled() { return _dynamicClassUnloadingEnabled; };
+	void unloadDeadClassLoaders(MM_EnvironmentBase *env);
+
+	void reportClassUnloadingStart(MM_EnvironmentBase *env);
+	void reportClassUnloadingEnd(MM_EnvironmentBase *env);
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
```

Comment:
Next cleanup: closing "endif" and opening "if" can be removed to coalesce two sections together